python/cuda_parallel/cuda/parallel/experimental/_cccl.py

```diff
@@ -0,0 +1,209 @@
+        ("single_tile_second_kernel", ctypes.c_void_p),
+        ("reduction_kernel", ctypes.c_void_p),
+    ]
+
+
+# MUST match `cccl_value_t` in c/include/cccl/c/types.h
+class Value(ctypes.Structure):
+    _fields_ = [("type", TypeInfo), ("state", ctypes.c_void_p)]
+
+
+def _type_to_enum(numba_type):
+    mapping = {
+        types.int8: TypeEnum.INT8,
+        types.int16: TypeEnum.INT16,
+        types.int32: TypeEnum.INT32,
+        types.int64: TypeEnum.INT64,
+        types.uint8: TypeEnum.UINT8,
+        types.uint16: TypeEnum.UINT16,
+        types.uint32: TypeEnum.UINT32,
+        types.uint64: TypeEnum.UINT64,
+        types.float32: TypeEnum.FLOAT32,
+        types.float64: TypeEnum.FLOAT64,
+    }
+    if numba_type in mapping:
+        return mapping[numba_type]
+    return TypeEnum.STORAGE
+
+
+# TODO: replace with functools.cache once our docs build environment
+# is upgraded to at least Python 3.9
+@functools.lru_cache(maxsize=None)
+def _numba_type_to_info(numba_type):
+    context = cuda.descriptor.cuda_target.target_context
+    value_type = context.get_value_type(numba_type)
+    size = value_type.get_abi_size(context.target_data)
+    alignment = value_type.get_abi_alignment(context.target_data)
+    return TypeInfo(size, alignment, _type_to_enum(numba_type))
+
+
+@functools.lru_cache(maxsize=None)
+def _numpy_type_to_info(numpy_type):
+    numba_type = numba.from_dtype(numpy_type)
+    return _numba_type_to_info(numba_type)
+
+
+def _device_array_to_cccl_iter(array):
+    dtype = array.dtype
+    info = _numpy_type_to_info(dtype)
+    # Note: this is slightly slower, but supports all ndarray-like objects as long as they support CAI
+    # TODO: switch to use gpumemoryview once it's ready
+    return Iterator(
+        info.size,
+        info.alignment,
+        IteratorKind.POINTER,
+        Op(),
+        Op(),
+        info,
+        array.__cuda_array_interface__["data"][0],
+    )
+
+
+def _iterator_to_cccl_iter(it):
+    context = cuda.descriptor.cuda_target.target_context
+    numba_type = it.numba_type
+    size = context.get_value_type(numba_type).get_abi_size(context.target_data)
+    alignment = context.get_value_type(numba_type).get_abi_alignment(
+        context.target_data
+    )
+    (advance_abi_name, advance_ltoir), (deref_abi_name, deref_ltoir) = it.ltoirs.items()
+    advance_op = Op(
+        OpKind.STATELESS,
+        advance_abi_name.encode("utf-8"),
+        ctypes.c_char_p(advance_ltoir),
+        len(advance_ltoir),
+        1,
+        1,
+        None,
+    )
+    deref_op = Op(
+        OpKind.STATELESS,
+        deref_abi_name.encode("utf-8"),
+        ctypes.c_char_p(deref_ltoir),
+        len(deref_ltoir),
+        1,
+        1,
+        None,
+    )
+    return Iterator(
+        size,
+        alignment,
+        OpKind.STATEFUL,
+        advance_op,
+        deref_op,
+        _numba_type_to_info(it.value_type),
+        it.state,
+    )
+
+
+def type_enum_as_name(enum_value):
```

Comment:
I'd move this up to be together with `_type_to_enum()`. (Because these functions are very closely related.)

Reply:
In this particular module, we have tons of private names and a handful of public names. I was inclined to keep the public APIs together at the end of the module. I don't think there's any an established convention though so let me know if you feel strongly.